### PR TITLE
Enable C++17 filesystem in the future

### PR DESCRIPTION
**Pull request by cybaol**
Idea 3 of #5881

### CODE REVIEW

**Comment by mvieth** [comment]
Thank you, but let us find a way to divide this into smaller, more manageable pull requests, please :smile: 
Firstly, let's exclude all files where something from `filesystem` is used as a function parameter (e.g. `boost::filesystem::path`) for now. This would be potentially changing the API/ABI, so we have to be very cautious there. We can think about that later. It seems like mostly tutorials and outofcore are affected.
Then split the rest in maybe two pull requests.
And finally, in the first pull request, switch the Ubuntu 23.04 to C++17 so we can be sure that the new code works correctly. Adding `CMAKE_ARGS: '-DCMAKE_CXX_STANDARD=17'` in https://github.com/PointCloudLibrary/pcl/blob/master/.ci/azure-pipelines/azure-pipelines.yaml#L56 should do the trick.
Thanks again!

**Comment by cybaol** [comment]
1. The function parameter `pcl_fs::path`  declared in `doc/tutorials` only appears in `.cpp` files, not `.h`  files.
2. Function `void getFilesInDirectory(bf::path & dir, std::string & rel_path_so_far, std::vector<std::string> & relative_paths, std::string & ext)` defined in `recognition` is a private member.

Is it possible for both 1 and 2 to change the API or ABI?

**Comment by cybaol** [comment]
For the `outofcore` module, if we want to replace `boost::filesystem::path` with `std::filesystem::path`, it will inevitably cause API break. But since the `outofcore` module hasn't been maintained in many years, we can do a minor code refactoring.
For example, in order to avoid unnecessary troubles during future maintenance, consider turning the function parameter `boost::filesystem::path` into `std::string`.

**Comment by mvieth** [comment]
You are right, the changes in `doc/tutorials` are not a problem. We can do those in a separate pull request.
The changes in `recognition` will not change the API (because the function is private, like you said), but it may change the ABI, at least when compiled with GCC or Clang. So we can make the changes in `recognition` after PCL 1.14.1 is released (we want no API/ABI break between 1.14.0 and 1.14.1, but an ABI break between 1.14.1 and 1.15.0 is okay). If you like, you can already create a separate pull request for the changes in `recognition`, and we can merge it after PCL 1.14.1 is released.

**Comment by mvieth** [comment]
Please undo the changes in `apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/pc_source/registered_views_source.h` and `apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/pc_source/source.h`.
`boost::filesystem::path`/`pcl_fs::path` is used as a function parameter in those files, and the files are installed (thus available to users). I will first have to consider what we do there. The changes in the other files are fine, so we can already merge those.

File: apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/pc_source/mesh_source.h [hunk 1]
```diff
@@ -231,7 +231,7 @@ class MeshSource : public Source<PointInT> {
     std::vector<std::string> files;
     std::string start;
     std::string ext = std::string("ply");
-    bf::path dir = path_;
+    pcl_fs::path dir = path_;
```
Comment (mvieth) [inline apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/pc_source/mesh_source.h:234]:
```suggestion
    bf::path dir = path_;
```
The function from apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/pc_source/source.h  is used. With `pcl_fs` there is an error: `cannot convert 'std::filesystem::path' to 'boost::filesystem::path&'`

**Review by mvieth — APPROVED** [review]
Thank you!

File: apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/utils/persistence_utils.h [hunk 2]
```diff
@@ -1,7 +1,14 @@
 #include <pcl/io/pcd_io.h>
 
 #include <boost/algorithm/string.hpp>
+
+#if (__cplusplus >= 201703L)
```
[the remark this answers is not in the document]
Comment (mvieth) [reply apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/utils/persistence_utils.h:5]: You are right, that would be better. Feel free to open a pull request

[the remark this answers is not in the document]
Comment (mvieth) [reply apps/3d_rec_framework/include/pcl/apps/3d_rec_framework/utils/persistence_utils.h:5]: https://github.com/PointCloudLibrary/pcl/pull/5945

**Pull request merged**